Audit logging over internal RPC

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR refactors `v/security/audit` to support producing to the audit topic by internal RPC. 

Event transport mode is selectable by a new cluster config `audit_use_rpc, ON by default.

This change needs load testing at scale, particularly to compare performance between rpc and kafka client modes. I made an effort to rework existing audit log scale tests, but the latency results we get from kgo-repeater don't seem particularly stable. Punting further analysis to a followup.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Improvements

* Introduce support for audit logging over internal RPC.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman commented:
/ci-repeat 1
release
skip-units
tests/rptest/scale_tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
skip-units
tests/rptest/scale_tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
skip-units
tests/rptest/scale_tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
skip-units
tests/rptest/scale_tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
skip-units
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69774-->
<!--743ae4bc-9df7-4dea-b7b3-d69b1895b8f1-->
<!--title-->
## Retry command for [Build#69774](https://buildkite.com/redpanda/redpanda/builds/69774)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestNoSecurityConfigured.test_bypass@{"audit_transport_mode":"kclient"}
tests/rptest/tests/audit_log_test.py::AuditLogTestNoSecurityConfigured.test_bypass@{"audit_transport_mode":"rpc"}
tests/rptest/tests/audit_log_test.py::AuditLogTestSmallBuffers.test_bypass@{"audit_transport_mode":"kclient"}
tests/rptest/tests/audit_log_test.py::AuditLogTestSmallBuffers.test_bypass@{"audit_transport_mode":"rpc"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4a22c742-1f2a-4882-9398-75cd04c77570-->
<!--TEST RESULTS-->
<!--d76d1c99-bdc0-47fd-b84e-4bfc450a293d-->
<!--TEST RESULTS-->
<!--784f8e69-da5d-425c-ae4d-2b49968a8bc7-->
<!--TEST RESULTS-->
<!--67ad3724-3772-4929-ac3a-a9cac45d2883-->
<!--TEST RESULTS-->
<!--407e59cb-13a4-4b7d-a0b9-641cb3be6168-->
<!--TEST RESULTS-->
<!--2e8e36ea-c0d3-4a99-84fe-3da4a5f4a606-->
<!--TEST RESULTS-->
<!--4c6f5bf7-ced4-43f4-930d-68c4b3c34e2d-->
<!--TEST RESULTS-->
<!--522929ba-d58e-47dc-9aab-420ae934d0fd-->
<!--TEST RESULTS-->
<!--417c08c0-4517-4d8d-b45d-157329741ab8-->
<!--TEST RESULTS-->
<!--b301e5fe-6613-435c-9e04-99438991f20a-->
<!--TEST RESULTS-->
<!--e9b32c1d-2c8b-4bff-8e5f-b8799addbe41-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69774">build#69774</a></summary>


| test_class                       | test_method   | test_arguments                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:---------------------------------|:--------------|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| AuditLogTestNoSecurityConfigured | test_bypass   | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/69774#0198522f-f3a8-466b-a56a-e34a22745e13 | FAIL          | 0/21     | The test has failed across all retries |
| AuditLogTestNoSecurityConfigured | test_bypass   | {"audit_transport_mode": "rpc"}     | integration | https://buildkite.com/redpanda/redpanda/builds/69774#0198522f-f3a8-466b-a56a-e34a22745e13 | FAIL          | 0/21     | The test has failed across all retries |
| AuditLogTestSmallBuffers         | test_bypass   | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/69774#0198522f-f3a8-466b-a56a-e34a22745e13 | FAIL          | 0/21     | The test has failed across all retries |
| AuditLogTestSmallBuffers         | test_bypass   | {"audit_transport_mode": "rpc"}     | integration | https://buildkite.com/redpanda/redpanda/builds/69774#0198522f-f3a8-466b-a56a-e34a22745e13 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69791">build#69791</a></summary>


| test_class   | test_method    | test_arguments          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:-------------|:---------------|:------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| AuditLogTest | test_audit_log | {"audit_use_rpc": true} | integration | https://buildkite.com/redpanda/redpanda/builds/69791#019852f9-dcf2-427d-8939-55907b09efcb | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69794">build#69794</a></summary>


| test_class   | test_method    | test_arguments          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    |
|:-------------|:---------------|:------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|
| AuditLogTest | test_audit_log | {"audit_use_rpc": true} | integration | https://buildkite.com/redpanda/redpanda/builds/69794#01985325-8870-42f7-8d2a-863b07a0a1cd | FLAKY         | 9/21     | The test was found to be new, and no failures are allowed |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70510">build#70510</a></summary>


| test_class              | test_method       | test_arguments                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:------------------------|:------------------|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| AuditLogTestEscapeHatch | test_escape_hatch | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/70510#019890a7-0eef-4470-9564-9bacea3622c6 | FAIL          | 0/23     | The test has failed across all retries |
| AuditLogTestEscapeHatch | test_escape_hatch | {"audit_transport_mode": "kclient"} | integration | https://buildkite.com/redpanda/redpanda/builds/70510#019890a9-e3b4-451c-8cfb-4b73413da209 | FAIL          | 0/23     | The test has failed across all retries |
| AuditLogTestEscapeHatch | test_escape_hatch | {"audit_transport_mode": "rpc"}     | integration | https://buildkite.com/redpanda/redpanda/builds/70510#019890a7-0eef-4470-9564-9bacea3622c6 | FAIL          | 0/23     | The test has failed across all retries |
| AuditLogTestEscapeHatch | test_escape_hatch | {"audit_transport_mode": "rpc"}     | integration | https://buildkite.com/redpanda/redpanda/builds/70510#019890a9-e3b4-451c-8cfb-4b73413da209 | FAIL          | 0/23     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70902">build#70902</a></summary>


| test_class   | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    |
|:-------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70902#0198b17e-5a61-49b0-84f1-25cb57b71d6e | FLAKY         | 16/25    | The test was found to be new, and no failures are allowed |
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70902#0198b17e-5a61-4d78-b59a-f34e29a112e2 | FLAKY         | 19/25    | The test was found to be new, and no failures are allowed |
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70902#0198b17e-5a62-44be-877e-45ad19b23f6e | FLAKY         | 18/25    | The test was found to be new, and no failures are allowed |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70903">build#70903</a></summary>


| test_class   | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    |
|:-------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70903#0198b1c0-f4d7-4507-bc5f-2abaeb59f2f6 | FLAKY         | 14/23    | The test was found to be new, and no failures are allowed |
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70903#0198b1c0-f4d7-4786-879b-804025154e2c | FLAKY         | 16/23    | The test was found to be new, and no failures are allowed |
| AuditLogTest | test_rpc_vs_kafka | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70903#0198b1c0-f4d7-4c89-8b88-eab00ffac5f7 | FLAKY         | 12/23    | The test was found to be new, and no failures are allowed |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70904">build#70904</a></summary>


| test_class             | test_method          | test_arguments                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:-----------------------|:---------------------|:-----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": false, "feature": 0, "install_license": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b469-807a-4fdf-84e2-6f149477f296 | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": false, "feature": 0, "install_license": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b48d-6a77-489e-be4b-be45f9f7a032 | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": true, "feature": 0, "install_license": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b469-807a-47db-9621-0d070ca13cc7 | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": true, "feature": 0, "install_license": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b48d-6a78-4472-9dcf-8eeb32f79648 | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": false, "feature": 0, "install_license": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b469-807b-4501-99cc-9a4eccf2d96d | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": false, "feature": 0, "install_license": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b48d-6a79-41c3-952f-acdb5b82d99d | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": true, "feature": 0, "install_license": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b469-807b-495d-b1dc-90978c1c9ce8 | FAIL          | 0/21     | The test has failed across all retries |
| EnterpriseFeaturesTest | test_enable_features | {"disable_trial": true, "feature": 0, "install_license": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/70904#0198b48d-6a79-4034-b6c7-68756704e5cc | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70915">build#70915</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70915#0198b984-4927-48fe-89d4-a33636e7c64d | FLAKY         | 18/21    | upstream reliability is '93.06282722513089'. current run reliability is '85.71428571428571'. drift is 7.34854 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71131">build#71131</a></summary>


| test_class       | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/71131#0198cb10-c6f1-4e16-b59f-f6406adabb4c | FLAKY         | 20/21    | upstream reliability is '98.24945295404814'. current run reliability is '95.23809523809523'. drift is 3.01136 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71867">build#71867</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71867#01992b88-cdad-468c-88bf-3aaf6ede60fa | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72050">build#72050</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72050#019936de-82ac-4f56-a277-a6d79204637a | FLAKY         | 20/21    | upstream reliability is '98.4536082474227'. current run reliability is '95.23809523809523'. drift is 3.21551 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/72050#019936de-82a4-4239-89c0-7413fa6cf998 | FLAKY         | 20/21    | upstream reliability is '98.38709677419355'. current run reliability is '95.23809523809523'. drift is 3.149 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69791-->
<!--62132dbf-02bd-4fff-bf97-9821e31e407c-->
<!--title-->
## Retry command for [Build#69791](https://buildkite.com/redpanda/redpanda/builds/69791)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_audit_log@{"audit_use_rpc":true}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69793-->
<!--9886f482-bc4b-490a-b166-abededc64ad6-->
<!--title-->
## Retry command for [Build#69793](https://buildkite.com/redpanda/redpanda/builds/69793)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_audit_log@{"audit_use_rpc":true}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69794-->
<!--1368a1b4-d4fb-43e3-8fd3-aa7d9c6dca16-->
<!--title-->
## Retry command for [Build#69794](https://buildkite.com/redpanda/redpanda/builds/69794)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_audit_log@{"audit_use_rpc":true}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
release
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
tests/rptest/scale_tests/audit_log_test.py
tests/rptest/tests/audit_log_test.py

oleiman commented:
/ci-repeat 1
release
tests/rptest/tests/audit_log_test.py

oleiman committed:
s/a: Forward declarations for audit logging

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/a: Wire kafka/data RPC client into audit_log_manager

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
config: Introduce audit_use_rpc

Does what it says on the box. Boolean defaults to true. Requires restart.

TODO: Bit of refactoring to relax restart requirement
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/a/tests: Introduce audit_test_utils

Extract some helper functions from audit_log_test.cc

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/a: Extract audit client from manager

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/a/client: Add rpc client & sink

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/a/mgr: Make audit log transport configurable by cluster config

Also wires audit_use_rpc into test fixtures and gtestize audit_log_test to
parametrize for rpc vs kclient.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/audit: Run ducktape tests in rpc and kafka mode both

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1
tests/rptest/tests/audit_log_test.py
dt-repeat=3

oleiman commented:
/ci-repeat 1
tests/rptest/tests/audit_log_test.py
dt-repeat=3

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70510-->
<!--4a49ee79-ce60-4284-b044-e7634da5ab13-->
<!--6a268411-a46c-4488-94d2-1c05ccd80bc4-->
<!--title-->
## Retry command for [Build#70510](https://buildkite.com/redpanda/redpanda/builds/70510)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestEscapeHatch.test_escape_hatch@{"audit_transport_mode":"kclient"}
tests/rptest/tests/audit_log_test.py::AuditLogTestEscapeHatch.test_escape_hatch@{"audit_transport_mode":"rpc"}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
tests/rptest/tests/audit_log_test.py
dt-repeat=1

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -1972,6 +1972,12 @@ configuration::configuration()
+  , audit_use_rpc(
+      *this,
+      "audit_use_rpc",
+      "If set, audit log will produce messages via internal redpanda RPCs",
```
```suggestion
      "If set, the audit log will produce messages via internal Redpanda RPCs.",
```

oleiman replied:
done. thanks!

oleiman commented:
/ci-repeat 1
release
tests/rptest/tests/audit_log_test.py
tests/rptest/scale_tests/audit_log_test.py

oleiman commented:
/ci-repeat 3
release
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_rpc_vs_kafka
dt-repeat 10

oleiman commented:
/ci-repeat 3
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_rpc_vs_kafka
dt-repeat=5

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70902-->
<!--f46b3795-c555-4a11-9e06-c464773f0e33-->
<!--title-->
## Retry command for [Build#70902](https://buildkite.com/redpanda/redpanda/builds/70902)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_rpc_vs_kafka
```
<!--/body-->

oleiman commented:
/ci-repeat 3
release
skip-redpanda-build
skip-units
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_rpc_vs_kafka
dt-repeat=3

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70903-->
<!--42ba4281-688a-438a-b88c-fc1622b83eae-->
<!--1a9813ca-2d9a-422b-9130-ff331a808806-->
<!--title-->
## Retry command for [Build#70903](https://buildkite.com/redpanda/redpanda/builds/70903)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/scale_tests/audit_log_test.py::AuditLogTest.test_rpc_vs_kafka
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70904-->
<!--f1ad57ee-5d99-404e-888d-ed722360a764-->
<!--71c77c5b-799b-48a3-af83-0f0db65df191-->
<!--d4e08abc-37ee-4973-95d3-f3e2748163f2-->
<!--b339d3ad-6291-468c-a224-758b81d3feec-->
<!--444794dd-8a18-48aa-9244-dadd9fe26b73-->
<!--fc8050dd-4bd7-4934-abdc-af149ce6d35c-->
<!--82b10ae8-68a7-477c-8a91-dd2982de602f-->
<!--d45fcb0c-32a9-404f-b20a-5daad399daa1-->
<!--title-->
## Retry command for [Build#70904](https://buildkite.com/redpanda/redpanda/builds/70904)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":true,"feature":0,"install_license":true}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":true,"feature":0,"install_license":false}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":false,"feature":0,"install_license":false}
tests/rptest/tests/enterprise_features_license_test.py::EnterpriseFeaturesTest.test_enable_features@{"disable_trial":false,"feature":0,"install_license":true}
```
<!--/body-->

oleiman commented on src/v/security/audit/client.cc:
```diff
@@ -28,6 +29,151 @@ namespace security::audit {
+
+        cluster::topic_properties audit_topic_props;
+        audit_topic_props.retention_bytes = tristate<size_t>{};
+        // TODO(oren): respect the replication config, otherwise it uses the
+        // default
```
pretty sure @michael-redpanda added support for this is a recent PR (or a pending one?)

oleiman replied:
it's here: https://github.com/redpanda-data/redpanda/pull/27245/commits/51dacee13df34d4e7499caff3b91d55a86c7614d

oleiman commented on src/v/security/audit/client.cc:
```diff
@@ -28,6 +29,151 @@ namespace security::audit {
+      model::partition_id pid,
+      audit_probe& probe,
+      ss::timer<>::time_point timeout) final {
+        // TODO(oren): map cluster errc to kafka. we should be able to do that
+        // actually with better fidelity. but really this is mostly for
+        // detecting auth errors I think, which is totally unnecessary here.
```
I'm thinking 'do nothing'

rockwotj replied:
This is fine, if you want this to be better the audit sink should have it's own `errc`

oleiman commented on src/v/security/audit/client.cc:
```diff
@@ -28,6 +29,151 @@ namespace security::audit {
+        return ss::now();
+    }
+    ss::future<> create_internal_topic() {
+        // constexpr std::string_view seven_days = "604800000";
```
cleanup

oleiman commented on src/v/security/audit/audit_log_manager.cc:
```diff
@@ -394,6 +406,7 @@ audit_log_manager::should_enqueue_audit_event() const {
         _probe.audit_error();
         return std::make_optional(audit_event_passthrough::yes);
     }
+    // TODO(oren): this is no-op if we're using an rpc client
```
```suggestion
    // NOTE: RPC client will never set this flag.
```

oleiman commented on src/v/security/audit/tests/audit_log_test.cc:
```diff
@@ -124,26 +132,28 @@ FIXTURE_TEST(test_audit_init_phase, kafka_client_fixture) {
     BOOST_CHECK(enqueued);
     BOOST_CHECK_EQUAL(
-      pending_audit_events(audit_mgr.local()).get(), number_events);
+      sa::test::pending_audit_events(audit_mgr.local()).get(), number_events);
```
these namespace changes should go in the utils config

oleiman commented on src/v/security/audit/tests/audit_log_test.cc:
```diff
@@ -124,26 +132,28 @@ FIXTURE_TEST(test_audit_init_phase, kafka_client_fixture) {
         config::shard_local_cfg().get("audit_enabled").set_value(true);
         /// Lower the fiber loop interval from 60s (set high so that messages
-        /// wouldn't be sent quicker then they could be enqueued) to a smaller
+        /// wouldn't be sent quicker then they could be enqueued) to a
+        /// smaller
```
formatting

oleiman commented on src/v/security/audit/tests/audit_log_rpc_test.cc:
```diff
@@ -0,0 +1,172 @@
+          .map_reduce0(std::move(enqueue_some), true, std::logical_and<>())
+          .get();
+
+    /// Since different messages related to application lifecycle may be
+    /// enqueued during program execution, the test solely asserts that at
+    ///    any
+    /// given time "if enough memory reservation does or does not exist,
+    /// should
+    /// the next enqueue work or not". Success is determined if the
+    ///    expectation
+    /// matches the observed outcome, on all attempts, across all shards.
```
formatting

oleiman commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1751,6 +1780,7 @@ def __init__(self, test_context):
                                                             self.password,
                                                             self.algorithm)))
 
+    # TODO(oren): maybe check that it passes in RPC mode
```
do it or remove

oleiman commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1794,6 +1824,7 @@ def __init__(self, test_context):
               self).__init__(test_context=test_context,
                              security=self._security_config)
 
+    # TODO(oren): should pass in RPC mode
```
do it or remove

oleiman commented:
/ci-repeat 1

oleiman commented:
@michael-redpanda @rockwotj - There's a bit more polish to be had here, but I think worth a round of review (low-priority of course) so we're all on the same page. TIA.

rockwotj commented on src/v/security/audit/fwd.h:
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
![what-year-is-it](https://github.com/user-attachments/assets/e2ce72d4-46c6-45fd-8ee8-6756f8783e0d)

oleiman replied:
![in-the-year-2000](https://github.com/user-attachments/assets/74888c54-2701-4f86-9a51-1dd859a49cd9)

michael-redpanda replied:
@oleiman that's some nostalgic stuff right there

oleiman replied:
©️ 2025 Redpanda Data, Inc

rockwotj commented on src/v/redpanda/application.cc:
```diff
@@ -1981,8 +1981,10 @@ void application::wire_up_redpanda_services(
-      &metadata_cache)
+      &metadata_cache,
+      &_kafka_data_rpc_client,
+      _audit_log_client_config.has_value() ? &_audit_log_client_config.value()
```
but this is always there right?

oleiman replied:
standard startup, yes, it's always there. but most (all?) of the whole cluster fixtures can bypass its initialization logic. can you spot the UB in this picture? 

https://github.com/redpanda-data/redpanda/blob/e4e1c9a7d9605cba220da6f307bb0dfb0650d49a/src/v/redpanda/application.cc#L1984

All that notwithstanding, I meant to clean this up. Probably fine to construct a meaningless default config and let it fail. The nullopt-ness of `_audit_log_client_config` isn't meaningful afaict.

rockwotj commented on src/v/config/configuration.cc:
```diff
@@ -1972,6 +1972,12 @@ configuration::configuration()
+      "audit_use_rpc",
+      "If set, the audit log will produce messages via internal redpanda RPCs",
+      {.needs_restart = needs_restart::yes, .visibility = visibility::tunable},
+      true)
```
> defaults to true

:fire:

oleiman replied:
🤠

rockwotj commented on src/v/security/audit/audit_log_manager.h:
```diff
@@ -264,6 +263,11 @@ class audit_log_manager
     /// Returns true if the internal fibers are up
     bool is_effectively_enabled() const { return _effectively_enabled; }
 
+    // TODO(oren): can't be misconfigured in rpc mode
```
Can you elaborate here? Maybe this should be some callback passed into the sink when it's created?

oleiman replied:
good idea

oleiman replied:
done

rockwotj commented on src/v/security/audit/audit_log_manager.cc:
```diff
@@ -793,8 +121,12 @@ audit_log_manager::audit_log_manager(
-    if (ss::this_shard_id() == client_shard_id && _config != nullptr) {
-        _sink = std::make_unique<audit_sink>(this, controller, *_config);
+
+    if (_config != nullptr) {
```
why is this `if` here?

oleiman replied:
gone

rockwotj commented on src/v/security/audit/client.h:
```diff
@@ -0,0 +1,167 @@
+    /// Primitives for ensuring background work and toggling of switch w/ async
+    /// work occur in lock step
+    ss::gate _gate;
+    mutex _mutex{"audit_log_manager::mutex"};
```
needs an update

oleiman replied:
done

rockwotj commented on src/v/security/audit/client.cc:
```diff
@@ -17,17 +17,164 @@
+            controller()->internal_topic_replication());
+        vlog(
+          adtlog.debug,
+          "TODO(plumbing): Attempting to create internal topic "
```
:)

oleiman replied:
for posterity: leaving it until mike's changes land

michael-redpanda replied:
I think the topic creator has merged :)

oleiman replied:
ah, forgot to add this. will do this evening

oleiman replied:
done

rockwotj commented on src/v/security/audit/client.cc:
```diff
@@ -370,6 +517,7 @@ ss::future<> audit_client::configure() {
         /// client to refresh its metadata on a subsequent request.
         /// Explicitly set `client::config::retries` to its default value.
         /// We might want to make this tunable at some point.
+        // TODO(oren): do we need to configure retries on the rpc client?
```
is this the right place for this TODO?

oleiman replied:
nope

oleiman replied:
done

rockwotj reviewed (commented):
Looks lovely. Thank you for pushing on this. Some questions is really all

rockwotj reviewed (dismissed):
I mean LGTM, but would love @michael-redpanda to review as he's more familiar with auditing.

oleiman commented:
> I mean LGTM, but would love @michael-redpanda to review as he's more familiar with auditing.

yup yup. @michael-redpanda whenever you find time

oleiman commented:
/cdt
tests/rptest/tests/audit_log_test.py
dt-repeat=3

michael-redpanda commented on src/v/security/audit/client.h:
```diff
@@ -0,0 +1,172 @@
+    audit_sink& operator=(audit_sink&&) = delete;
+    virtual ~audit_sink();
+
+    /// Starts a kafka::client if none is allocated, backgrounds the work
```
nit - I don't think this comment is correct anymore

oleiman replied:
Ah, yeah `kafka::client`. 👌

oleiman replied:
done

michael-redpanda commented on src/v/security/audit/client.cc:
```diff
@@ -0,0 +1,663 @@
+            co_await publish_app_lifecycle_event(
+              application_lifecycle::activity_id::start);
+            co_await resume();
+            // co_await _audit_mgr->resume();
```
nit: commented out code

oleiman replied:
done

michael-redpanda commented on src/v/security/audit/client.cc:
```diff
@@ -0,0 +1,663 @@
+        co_await publish_app_lifecycle_event(
+          application_lifecycle::activity_id::stop);
+        co_await pause();
+        // co_await _audit_mgr->pause();
```
nit commented otu code

michael-redpanda commented on src/v/security/audit/client.cc:
```diff
@@ -17,17 +17,165 @@
+        }
+
+        // report unknown server error if we aborted before making any attempt
+        if (auto errc = ec.value_or(kafka::error_code::unknown_server_error);
```
should there be a special handling here for a timeout?

oleiman replied:
we didn't (and still don't) have anything special in the kafka::client version. anything particular in mind? backoff or something?

michael-redpanda replied:
I think the retries were baked into the `kafka::client` itself

oleiman replied:
oh yeah, there should be retries internal to the rpc client. iirc it will return an error if exhausted, then we should circle around and try again. i will take another look to confirm.

michael-redpanda commented on src/v/security/audit/client.cc:
```diff
@@ -0,0 +1,663 @@
+
+    void reset_client() final { _client.reset(nullptr); }
+
+    ss::future<> pause() final {
+        return manager()->container().invoke_on_all(
+          [](audit_log_manager& mgr) { return mgr.pause(); });
+    }
+    ss::future<> resume() final {
+        return manager()->container().invoke_on_all(
+          [](audit_log_manager& mgr) { return mgr.resume(); });
+    }
```
Is this correct?  `manager()` returns an `audit_log_mgr*` right?  I guess it's different from the rpc sink implementation

oleiman replied:
> `manager()` returns an `audit_log_manager*`

yeah, but `audit_log_manager` is derived from `peering_sharded_service`, which means (from core/sharded.hh):

```
/// If a service class inherits from this, it will gain a \code container()
/// \endcode method that provides access to the \ref sharded object, with which
/// it can call its peers.
```

The difference between the two implementations is that we have an active rpc client on every shard but only an active kafka client on shard 0. So pause/resume has to fan out to unlock the per-shard queues in the manager. That's the intent anyway. Does this sound right?

michael-redpanda reviewed (commented):
This looks amazing

oleiman commented:
putting in draft to rebase & push

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/ffa0146a85019cb3c28a3caec68b8a48d52334c2..e04f2a253c91ba823d2cafbb7c3f42dd80b11607) to resolve merge conflicts (mostly python linter related). CR changes in the next push.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/e04f2a253c91ba823d2cafbb7c3f42dd80b11607..cca716babb5a2c3ec06b790d2193f10111277217) address CR

oleiman committed:
k/d/rpc: Plumb replication factor into client::create_topic

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/cca716babb5a2c3ec06b790d2193f10111277217..4798c99ee21b5581588faba4251a34bef73ed378) respect configured replication factor for audit topic when creating through rpc client

dotnwat reviewed (commented):
an excellent bit of ascii

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -1982,6 +1982,12 @@ configuration::configuration()
+  , audit_use_rpc(
+      *this,
+      "audit_use_rpc",
+      "If set, the audit log will produce messages via internal redpanda RPCs",
```
If the default is true, I would flip the description. Review for tech accuracy and apply the linter: 

```suggestion
      "Produces audit log messages using internal Redpanda RPCs. When disabled, Redpanda uses Kafka client mode instead.",
```

oleiman replied:
fair point, will do.

oleiman replied:
some definition of done. lmk if it looks ok.

paulohtb6 replied:
nice

michael-redpanda reviewed (dismissed):
🔥 LFG 🔥

michael-redpanda reviewed (dismissed):
🔥 LFG 🔥 
(again)

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/4798c99ee21b5581588faba4251a34bef73ed378..015615a9e904cd19a5966761f7ebd8347a4aa71c) to fix config doc

[force push](https://github.com/redpanda-data/redpanda/compare/015615a9e904cd19a5966761f7ebd8347a4aa71c..e205626a08459870c8d43f61291162ef794315a2) to rebase & fix merge collision

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72191-->
<!--33dac995-1759-4bf5-9f23-4900ed38779b-->
<!--e35a314e-6fb1-4237-ab97-bb311ef86285-->
<!--6be7bd9d-90f7-4f64-b268-0f693e58836d-->
<!--3a37ef55-05be-470b-82db-cd98ce386e77-->
<!--title-->
## Retry command for [Build#72191](https://buildkite.com/redpanda/redpanda/builds/72191)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestOauth.test_kafka_oauth@{"audit_transport_mode":"rpc","authz_match":"rbac"}
tests/rptest/tests/audit_log_test.py::AuditLogTestOauth.test_kafka_oauth@{"audit_transport_mode":"kclient","authz_match":"acl"}
tests/rptest/tests/audit_log_test.py::AuditLogTestOauth.test_kafka_oauth@{"audit_transport_mode":"rpc","authz_match":"acl"}
tests/rptest/tests/audit_log_test.py::AuditLogTestOauth.test_kafka_oauth@{"audit_transport_mode":"kclient","authz_match":"rbac"}
```
<!--/body-->

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/e205626a08459870c8d43f61291162ef794315a2..584fa058d2c7cdf2253fed5691611e8176955d56) rebase dev for merge conflicts on formatting and seemingly oauth test fix

oleiman commented:
💥

michael-redpanda commented:
🔥 LFG 🔥
(again)x2
